--- a/include/soc/tegra/fuse.h
+++ b/include/soc/tegra/fuse.h
@@ -30,6 +30,7 @@ bool tegra_is_silicon(void);
 enum tegra_revision {
 	TEGRA_REVISION_UNKNOWN = 0,
 	TEGRA_REVISION_A01,
+	TEGRA_REVISION_A01q,
 	TEGRA_REVISION_A02,
 	TEGRA_REVISION_A03,
 	TEGRA_REVISION_A03p,

--- a/drivers/soc/tegra/fuse/speedo-tegra210.c
+++ b/drivers/soc/tegra/fuse/speedo-tegra210.c
@@ -58,6 +58,10 @@ static void __init rev_sku_to_speedo_ids
 					 u8 speedo_rev, int *threshold)
 {
 	int sku = sku_info->sku_id;
+	int rev = sku_info->revision;
+	bool a02 = (rev != TEGRA_REVISION_A01) && (rev != TEGRA_REVISION_A01q) &&
+		(rev != TEGRA_REVISION_UNKNOWN);
+	bool always_on = true; // false;
 
 	/* Assign to default */
 	sku_info->cpu_speedo_id = 0;
@@ -82,6 +86,15 @@ static void __init rev_sku_to_speedo_ids
 		sku_info->cpu_speedo_id = 1;
 		break;
 
+	case 0x8F:
+		if (a02 && always_on) {
+			sku_info->cpu_speedo_id = 9;
+			sku_info->gpu_speedo_id = 2;
+			// sku_info->ucm = TEGRA_UCM2;
+			break;
+		}
+		fallthrough;
+
 	default:
 		pr_err("Tegra210: unknown SKU %#04x\n", sku);
 		/* Using the default for the error case */
