--- a/drivers/pci/controller/pci-tegra.c
+++ b/drivers/pci/controller/pci-tegra.c
@@ -28,12 +28,16 @@
 #include <linux/module.h>
 #include <linux/msi.h>
 #include <linux/of_address.h>
+#include <linux/of_gpio.h>
 #include <linux/of_pci.h>
 #include <linux/of_platform.h>
 #include <linux/pci.h>
 #include <linux/phy/phy.h>
 #include <linux/pinctrl/consumer.h>
 #include <linux/platform_device.h>
+#if IS_ENABLED(CONFIG_TEGRA_BWMGR)
+#include <linux/platform/tegra/emc_bwmgr.h>
+#endif
 #include <linux/reset.h>
 #include <linux/sizes.h>
 #include <linux/slab.h>
@@ -47,6 +51,9 @@
 
 #define INT_PCI_MSI_NR (8 * 32)
 
+#define PCI_CFG_SPACE_SIZE	256
+#define PCI_EXT_CFG_SPACE_SIZE	4096
+
 /* register definitions */
 
 #define AFI_AXI_BAR0_SZ	0x00
@@ -163,10 +170,93 @@
 
 #define AFI_PLLE_CONTROL		0x160
 #define  AFI_PLLE_CONTROL_BYPASS_PADS2PLLE_CONTROL (1 << 9)
+#define  AFI_PLLE_CONTROL_BYPASS_PCIE2PLLE_CONTROL (1 << 8)
 #define  AFI_PLLE_CONTROL_PADS2PLLE_CONTROL_EN (1 << 1)
+#define  AFI_PLLE_CONTROL_PCIE2PLLE_CONTROL_EN (1 << 0)
 
 #define AFI_PEXBIAS_CTRL_0		0x168
 
+#define RP_INTR_BCR	0x3c
+#define  RP_INTR_BCR_INTR_LINE		(0xff << 0)
+#define  RP_INTR_BCR_SB_RESET		(0x1 << 22)
+
+#define RP_L1_PM_SUBSTATES_CTL		0xc00
+#define  RP_L1_PM_SUBSTATES_CTL_PCI_PM_L1_2		(0x1 << 0)
+#define  RP_L1_PM_SUBSTATES_CTL_PCI_PM_L1_1		(0x1 << 1)
+#define  RP_L1_PM_SUBSTATES_CTL_ASPM_L1_2		(0x1 << 2)
+#define  RP_L1_PM_SUBSTATES_CTL_ASPM_L1_1		(0x1 << 3)
+#define  RP_L1_PM_SUBSTATES_CTL_CM_RTIME_MASK		(0xff << 8)
+#define  RP_L1_PM_SUBSTATES_CTL_CM_RTIME_SHIFT		8
+#define  RP_L1_PM_SUBSTATES_CTL_T_PWRN_SCL_MASK		(0x3 << 16)
+#define  RP_L1_PM_SUBSTATES_CTL_T_PWRN_SCL_SHIFT	16
+#define  RP_L1_PM_SUBSTATES_CTL_T_PWRN_VAL_MASK		(0x1f << 19)
+#define  RP_L1_PM_SUBSTATES_CTL_T_PWRN_VAL_SHIFT	19
+#define  RP_L1_PM_SUBSTATES_CTL_HIDE_CAP		(0x1 << 24)
+
+#define RP_L1_PM_SUBSTATES_1_CTL	0xc04
+#define  RP_L1_PM_SUBSTATES_1_CTL_PWR_OFF_DLY_MASK	0x1fff
+#define  RP_L1_PM_SUBSTATES_1_CTL_PWR_OFF_DLY		0x26
+#define  RP_L1SS_1_CTL_CLKREQ_ASSERTED_DLY_MASK		(0x1ff << 13)
+#define  RP_L1SS_1_CTL_CLKREQ_ASSERTED_DLY		(0x27 << 13)
+
+#define RP_L1_PM_SUBSTATES_2_CTL	0xc08
+#define  RP_L1_PM_SUBSTATES_2_CTL_T_L1_2_DLY_MASK	0x1fff
+#define  RP_L1_PM_SUBSTATES_2_CTL_T_L1_2_DLY		0x4d
+#define  RP_L1_PM_SUBSTATES_2_CTL_MICROSECOND_MASK	(0xff << 13)
+#define  RP_L1_PM_SUBSTATES_2_CTL_MICROSECOND		(0x13 << 13)
+#define  RP_L1_PM_SUBSTATES_2_CTL_MICROSECOND_COMP_MASK	(0xf << 21)
+#define  RP_L1_PM_SUBSTATES_2_CTL_MICROSECOND_COMP	(0x2 << 21)
+
+#define RP_LTR_REP_VAL			0xc10
+
+#define RP_L1_1_ENTRY_COUNT		0xc14
+#define  RP_L1_1_ENTRY_COUNT_RESET	(1 << 31)
+
+#define RP_L1_2_ENTRY_COUNT		0xc18
+#define  RP_L1_2_ENTRY_COUNT_RESET	(1 << 31)
+
+#define RP_TIMEOUT0		0xe24
+#define  RP_TIMEOUT0_PAD_PWRUP_MASK		0xff
+#define  RP_TIMEOUT0_PAD_PWRUP			0xa
+#define  RP_TIMEOUT0_PAD_PWRUP_CM_MASK		0xffff00
+#define  RP_TIMEOUT0_PAD_PWRUP_CM		(0x180 << 8)
+#define  RP_TIMEOUT0_PAD_SPDCHNG_GEN2_MASK	(0xff << 24)
+#define  RP_TIMEOUT0_PAD_SPDCHNG_GEN2		(0xa << 24)
+
+#define RP_TIMEOUT1		0xe28
+#define  RP_TIMEOUT1_RCVRY_SPD_SUCCESS_EIDLE_MASK	(0xff << 16)
+#define  RP_TIMEOUT1_RCVRY_SPD_SUCCESS_EIDLE		(0x10 << 16)
+#define  RP_TIMEOUT1_RCVRY_SPD_UNSUCCESS_EIDLE_MASK	(0xff << 24)
+#define  RP_TIMEOUT1_RCVRY_SPD_UNSUCCESS_EIDLE		(0x74 << 24)
+
+#define RP_PRBS			0xe34
+#define  RP_PRBS_LOCKED			(1 << 16)
+
+#define RP_LANE_PRBS_ERR_COUNT	0xe38
+
+#define RP_LTSSM_DBGREG		0xe44
+#define  RP_LTSSM_DBGREG_LINKFSM16	(1 << 16)
+
+#define RP_LTSSM_TRACE_CONTROL	0xe50
+#define  LTSSM_TRACE_CONTROL_CLEAR_STORE_EN			(1 << 0)
+#define  LTSSM_TRACE_CONTROL_CLEAR_RAM				(1 << 2)
+#define  LTSSM_TRACE_CONTROL_TRIG_ON_EVENT			(1 << 3)
+#define  LTSSM_TRACE_CONTROL_TRIG_LTSSM_MAJOR_OFFSET		4
+#define  LTSSM_TRACE_CONTROL_TRIG_PTX_LTSSM_MINOR_OFFSET	8
+#define  LTSSM_TRACE_CONTROL_TRIG_PRX_LTSSM_MAJOR_OFFSET	11
+
+#define RP_LTSSM_TRACE_STATUS	0xe54
+#define  LTSSM_TRACE_STATUS_PRX_MINOR(reg)		(((reg) >> 19) & 0x7)
+#define  LTSSM_TRACE_STATUS_PTX_MINOR(reg)		(((reg) >> 16) & 0x7)
+#define  LTSSM_TRACE_STATUS_MAJOR(reg)			(((reg) >> 12) & 0xf)
+#define  LTSSM_TRACE_STATUS_READ_DATA_VALID(reg)	(((reg) >> 11) & 0x1)
+#define  LTSSM_TRACE_STATUS_READ_ADDR(reg)		((reg) << 6)
+#define  LTSSM_TRACE_STATUS_WRITE_POINTER(reg)		(((reg) >> 1) & 0x1f)
+#define  LTSSM_TRACE_STATUS_RAM_FULL(reg)		(reg & 0x1)
+
+#define RP_ECTL_1_R1	0x00000e80
+#define  RP_ECTL_1_R1_TX_DRV_AMP_1C_MASK	0x3f
+
 #define RP_ECTL_2_R1	0x00000e84
 #define  RP_ECTL_2_R1_RX_CTLE_1C_MASK		0xffff
 
@@ -180,6 +270,9 @@
 #define RP_ECTL_6_R1	0x00000e94
 #define  RP_ECTL_6_R1_RX_EQ_CTRL_H_1C_MASK	0xffffffff
 
+#define RP_ECTL_1_R2	0x00000ea0
+#define  RP_ECTL_1_R2_TX_DRV_AMP_1C_MASK	0x3f
+
 #define RP_ECTL_2_R2	0x00000ea4
 #define  RP_ECTL_2_R2_RX_CTLE_1C_MASK	0xffff
 
@@ -198,6 +291,22 @@
 #define  RP_VEND_XP_OPPORTUNISTIC_ACK		(1 << 27)
 #define  RP_VEND_XP_OPPORTUNISTIC_UPDATEFC	(1 << 28)
 #define  RP_VEND_XP_UPDATE_FC_THRESHOLD_MASK	(0xff << 18)
+#define  RP_VEND_XP_PRBS_STAT			(0xffff << 2)
+#define  RP_VEND_XP_PRBS_EN			(1 << 1)
+
+#define RP_VEND_XP1	0xf04
+#define  RP_VEND_XP1_LINK_PVT_CTL_IGNORE_L0S		(1 << 23)
+#define  RP_VEND_XP1_LINK_PVT_CTL_L1_ASPM_SUPPORT	(1 << 21)
+#define  RP_VEND_XP1_RNCTRL_MAXWIDTH_MASK		(0x3f << 0)
+#define  RP_VEND_XP1_RNCTRL_EN				(1 << 7)
+
+#define RP_XP_REF	0xf30
+#define  RP_XP_REF_MICROSECOND_LIMIT_MASK	0xff
+#define  RP_XP_REF_MICROSECOND_LIMIT		0x14
+#define  RP_XP_REF_MICROSECOND_ENABLE		(1 << 8)
+#define  RP_XP_REF_CPL_TO_OVERRIDE		(1 << 13)
+#define  RP_XP_REF_CPL_TO_CUSTOM_VALUE_MASK	(0x1ffff << 14)
+#define  RP_XP_REF_CPL_TO_CUSTOM_VALUE		(0x1770 << 14)
 
 #define RP_VEND_CTL0	0x00000f44
 #define  RP_VEND_CTL0_DSK_RST_PULSE_WIDTH_MASK	(0xf << 12)
@@ -212,6 +321,10 @@
 #define RP_VEND_CTL2 0x00000fa8
 #define  RP_VEND_CTL2_PCA_ENABLE (1 << 7)
 
+#define RP_PRIV_XP_CONFIG	0xfac
+#define  RP_PRIV_XP_CONFIG_LOW_PWR_DURATION_MASK	0x3
+#define  RP_PRIV_XP_DURATION_IN_LOW_PWR_100NS		0xfb0
+
 #define RP_PRIV_MISC	0x00000fe0
 #define  RP_PRIV_MISC_PRSNT_MAP_EP_PRSNT		(0xe << 0)
 #define  RP_PRIV_MISC_PRSNT_MAP_EP_ABSNT		(0xf << 0)
@@ -222,12 +335,38 @@
 #define  RP_PRIV_MISC_TMS_CLK_CLAMP_THRESHOLD		(0xf << 24)
 #define  RP_PRIV_MISC_TMS_CLK_CLAMP_ENABLE		(1 << 31)
 
+#define RP_XP_CTL_1	0xfec
+#define  RP_XP_CTL_1_OLD_IOBIST_EN	(1 << 25)
+
+#define RP_VEND_XP_PAD_PWRDN	0x00000f50
+#define  RP_VEND_XP_PAD_PWRDN_L1_EN			(1 << 0)
+#define  RP_VEND_XP_PAD_PWRDN_DYNAMIC_EN		(1 << 1)
+#define  RP_VEND_XP_PAD_PWRDN_DISABLED_EN		(1 << 2)
+#define  RP_VEND_XP_PAD_PWRDN_L1_CLKREQ_EN		(1 << 15)
+#define  RP_VEND_XP_PAD_PWRDN_SLEEP_MODE_DYNAMIC_L1PP	(3 << 5)
+#define  RP_VEND_XP_PAD_PWRDN_SLEEP_MODE_L1_L1PP	(3 << 3)
+#define  RP_VEND_XP_PAD_PWRDN_SLEEP_MODE_L1_CLKREQ_L1PP	(3 << 16)
+
+#define RP_PRIV_XP_RX_L0S_ENTRY_COUNT	0xf8C
+#define RP_PRIV_XP_TX_L0S_ENTRY_COUNT	0xf90
+#define RP_PRIV_XP_TX_L1_ENTRY_COUNT	0xf94
+
 #define RP_LINK_CONTROL_STATUS			0x00000090
 #define  RP_LINK_CONTROL_STATUS_DL_LINK_ACTIVE	0x20000000
 #define  RP_LINK_CONTROL_STATUS_LINKSTAT_MASK	0x3fff0000
+#define  RP_LINK_CONTROL_STATUS_NEG_LINK_WIDTH	(0x3f << 20)
+#define  RP_LINK_CONTROL_STATUS_LINK_SPEED	(0xf << 16)
+#define  RP_LINK_CONTROL_STATUS_L1_ENABLED	(1 << 1)
+#define  RP_LINK_CONTROL_STATUS_L0s_ENABLED	(1 << 0)
 
 #define RP_LINK_CONTROL_STATUS_2		0x000000b0
 
+#define RP_L1_PM_SUBSTATES_CAP	0x144
+
+#define RP_L1_PM_SS_CONTROL	0x148
+#define  RP_L1_PM_SS_CONTROL_ASPM_L11_ENABLE	0x8
+#define  RP_L1_PM_SS_CONTROL_ASPM_L12_ENABLE	0x4
+
 #define PADS_CTL_SEL		0x0000009c
 
 #define PADS_CTL		0x000000a0
@@ -266,6 +405,7 @@
 #define LINK_RETRAIN_TIMEOUT 100000 /* in usec */
 
 struct tegra_msi {
+	struct msi_controller chip;
 	DECLARE_BITMAP(used, INT_PCI_MSI_NR);
 	struct irq_domain *domain;
 	struct mutex map_lock;
@@ -283,6 +423,11 @@ struct tegra_pcie_port_soc {
 	} pme;
 };
 
+struct pcie_dvfs {
+	u32 afi_clk;
+	u32 emc_clk;
+};
+
 struct tegra_pcie_soc {
 	unsigned int num_ports;
 	const struct tegra_pcie_port_soc *ports;
@@ -304,12 +449,21 @@ struct tegra_pcie_soc {
 	bool program_deskew_time;
 	bool update_fc_timer;
 	bool has_cache_bars;
+	bool enable_wrap;
+	bool has_aspm_l1;
+	bool has_aspm_l1ss;
+	bool l1ss_rp_wake_fixup;
+	bool dvfs_mselect;
+	bool dvfs_afi;
+	struct pcie_dvfs dfs_tbl[10][2];
 	struct {
 		struct {
+			u32 rp_ectl_1_r1;
 			u32 rp_ectl_2_r1;
 			u32 rp_ectl_4_r1;
 			u32 rp_ectl_5_r1;
 			u32 rp_ectl_6_r1;
+			u32 rp_ectl_1_r2;
 			u32 rp_ectl_2_r2;
 			u32 rp_ectl_4_r2;
 			u32 rp_ectl_5_r2;
@@ -338,6 +492,10 @@ struct tegra_pcie {
 	struct reset_control *afi_rst;
 	struct reset_control *pcie_xrst;
 
+#if IS_ENABLED(CONFIG_TEGRA_BWMGR)
+	struct tegra_bwmgr_client *emc_bwmgr;
+#endif
+
 	bool legacy_phy;
 	struct phy *phy;
 
@@ -349,6 +507,8 @@ struct tegra_pcie {
 	struct regulator_bulk_data *supplies;
 	unsigned int num_supplies;
 
+	int pex_wake;
+
 	const struct tegra_pcie_soc *soc;
 	struct dentry *debugfs;
 };
@@ -366,12 +526,27 @@ struct tegra_pcie_port {
 	void __iomem *base;
 	unsigned int index;
 	unsigned int lanes;
+	unsigned int loopback_stat;
+	unsigned int aspm_state;
+	bool supports_clkreq;
+
+	int n_gpios;
+	int *gpios;
+	bool has_mxm_port;
+	int pwr_gd_gpio;
 
 	struct phy **phys;
 
 	struct gpio_desc *reset_gpio;
+	struct dentry *port_debugfs;
 };
 
+static bool is_gen2_speed;
+static u16 bdf;
+static u16 config_offset;
+static u32 config_val;
+static u16 config_aspm_state;
+
 static inline void afi_writel(struct tegra_pcie *pcie, u32 value,
 			      unsigned long offset)
 {
@@ -394,6 +569,14 @@ static inline u32 pads_readl(struct tegr
 	return readl(pcie->pads + offset);
 }
 
+static bool tegra_pcie_link_up(struct tegra_pcie_port *port)
+{
+	u32 value;
+
+	value = readl(port->base + RP_LINK_CONTROL_STATUS);
+	return !!(value & RP_LINK_CONTROL_STATUS_DL_LINK_ACTIVE);
+}
+
 /*
  * The configuration space mapping on Tegra is somewhat similar to the ECAM
  * defined by PCIe. However it deviates a bit in how the 4 bits for extended
@@ -459,20 +642,50 @@ static void __iomem *tegra_pcie_map_bus(
 static int tegra_pcie_config_read(struct pci_bus *bus, unsigned int devfn,
 				  int where, int size, u32 *value)
 {
+	struct tegra_pcie *pcie = bus->sysdata;
+	struct pci_dev *bridge;
+	struct tegra_pcie_port *port;
+
 	if (bus->number == 0)
 		return pci_generic_config_read32(bus, devfn, where, size,
 						 value);
 
+	bridge = pcie_find_root_port(bus->self);
+
+	list_for_each_entry(port, &pcie->ports, list)
+		if (port->index + 1 == PCI_SLOT(bridge->devfn))
+			break;
+
+	/* If there is no link, then there is no device */
+	if (!tegra_pcie_link_up(port)) {
+		*value = 0xffffffff;
+		return PCIBIOS_DEVICE_NOT_FOUND;
+	}
+
 	return pci_generic_config_read(bus, devfn, where, size, value);
 }
 
 static int tegra_pcie_config_write(struct pci_bus *bus, unsigned int devfn,
 				   int where, int size, u32 value)
 {
+	struct tegra_pcie *pcie = bus->sysdata;
+	struct tegra_pcie_port *port;
+	struct pci_dev *bridge;
+
 	if (bus->number == 0)
 		return pci_generic_config_write32(bus, devfn, where, size,
 						  value);
 
+	bridge = pcie_find_root_port(bus->self);
+
+	list_for_each_entry(port, &pcie->ports, list)
+		if (port->index + 1 == PCI_SLOT(bridge->devfn))
+			break;
+
+	/* If there is no link, then there is no device */
+	if (!tegra_pcie_link_up(port))
+		return PCIBIOS_DEVICE_NOT_FOUND;
+
 	return pci_generic_config_write(bus, devfn, where, size, value);
 }
 
@@ -529,6 +742,44 @@ static void tegra_pcie_port_reset(struct
 	}
 }
 
+static void disable_aspm_l0s(struct tegra_pcie_port *port)
+{
+	u32 val = 0;
+
+	val = readl(port->base + RP_VEND_XP1);
+	val |= RP_VEND_XP1_LINK_PVT_CTL_IGNORE_L0S;
+	writel(val, port->base + RP_VEND_XP1);
+}
+
+static void disable_aspm_l10(struct tegra_pcie_port *port)
+{
+	u32 val = 0;
+
+	val = readl(port->base + RP_VEND_XP1);
+	val &= ~RP_VEND_XP1_LINK_PVT_CTL_L1_ASPM_SUPPORT;
+	writel(val, port->base + RP_VEND_XP1);
+}
+
+static void disable_aspm_l11(struct tegra_pcie_port *port)
+{
+	u32 val = 0;
+
+	val = readl(port->base + RP_L1_PM_SUBSTATES_CTL);
+	val &= ~RP_L1_PM_SUBSTATES_CTL_PCI_PM_L1_1;
+	val &= ~RP_L1_PM_SUBSTATES_CTL_ASPM_L1_1;
+	writel(val, port->base + RP_L1_PM_SUBSTATES_CTL);
+}
+
+static void disable_aspm_l12(struct tegra_pcie_port *port)
+{
+	u32 val = 0;
+
+	val = readl(port->base + RP_L1_PM_SUBSTATES_CTL);
+	val &= ~RP_L1_PM_SUBSTATES_CTL_PCI_PM_L1_2;
+	val &= ~RP_L1_PM_SUBSTATES_CTL_ASPM_L1_2;
+	writel(val, port->base + RP_L1_PM_SUBSTATES_CTL);
+}
+
 static void tegra_pcie_enable_rp_features(struct tegra_pcie_port *port)
 {
 	const struct tegra_pcie_soc *soc = port->pcie->soc;
@@ -565,6 +816,47 @@ static void tegra_pcie_enable_rp_feature
 	}
 
 	writel(value, port->base + RP_PRIV_MISC);
+
+	if (soc->has_aspm_l1) {
+		/* Advertise ASPM-L1 state capability*/
+		value = readl(port->base + RP_VEND_XP1);
+		value |= RP_VEND_XP1_LINK_PVT_CTL_L1_ASPM_SUPPORT;
+		writel(value, port->base + RP_VEND_XP1);
+
+		/* Power saving configuration for L1 sleep/idle */
+		value = readl(port->base + RP_VEND_XP_PAD_PWRDN);
+		value |= RP_VEND_XP_PAD_PWRDN_DISABLED_EN;
+		value |= RP_VEND_XP_PAD_PWRDN_DYNAMIC_EN;
+		value |= RP_VEND_XP_PAD_PWRDN_L1_EN;
+		value |= RP_VEND_XP_PAD_PWRDN_L1_CLKREQ_EN;
+		value |= RP_VEND_XP_PAD_PWRDN_SLEEP_MODE_DYNAMIC_L1PP;
+		value |= RP_VEND_XP_PAD_PWRDN_SLEEP_MODE_L1_L1PP;
+		value |= RP_VEND_XP_PAD_PWRDN_SLEEP_MODE_L1_CLKREQ_L1PP;
+		writel(value, port->base + RP_VEND_XP_PAD_PWRDN);
+
+		if (port->aspm_state & 0x1)
+			disable_aspm_l0s(port);
+		if (port->aspm_state & 0x2)
+			disable_aspm_l10(port);
+	}
+
+	if (soc->has_aspm_l1ss) {
+		if (port->aspm_state & 0x2) {
+			disable_aspm_l11(port);
+			disable_aspm_l12(port);
+		}
+		if (port->aspm_state & 0x4)
+			disable_aspm_l11(port);
+		if (port->aspm_state & 0x8)
+			disable_aspm_l12(port);
+
+		/* Disable L1SS capability if CLKREQ# is not present */
+		if (!port->supports_clkreq) {
+			value = readl(port->base + RP_L1_PM_SUBSTATES_CTL);
+			value |= RP_L1_PM_SUBSTATES_CTL_HIDE_CAP;
+			writel(value, port->base + RP_L1_PM_SUBSTATES_CTL);
+		}
+	}
 }
 
 static void tegra_pcie_program_ectl_settings(struct tegra_pcie_port *port)
@@ -572,6 +864,11 @@ static void tegra_pcie_program_ectl_sett
 	const struct tegra_pcie_soc *soc = port->pcie->soc;
 	u32 value;
 
+	value = readl(port->base + RP_ECTL_1_R1);
+	value &= ~RP_ECTL_1_R1_TX_DRV_AMP_1C_MASK;
+	value |= soc->ectl.regs.rp_ectl_1_r1;
+	writel(value, port->base + RP_ECTL_1_R1);
+
 	value = readl(port->base + RP_ECTL_2_R1);
 	value &= ~RP_ECTL_2_R1_RX_CTLE_1C_MASK;
 	value |= soc->ectl.regs.rp_ectl_2_r1;
@@ -593,6 +890,11 @@ static void tegra_pcie_program_ectl_sett
 	value |= soc->ectl.regs.rp_ectl_6_r1;
 	writel(value, port->base + RP_ECTL_6_R1);
 
+	value = readl(port->base + RP_ECTL_1_R2);
+	value &= ~RP_ECTL_1_R2_TX_DRV_AMP_1C_MASK;
+	value |= soc->ectl.regs.rp_ectl_1_r2;
+	writel(value, port->base + RP_ECTL_1_R2);
+
 	value = readl(port->base + RP_ECTL_2_R2);
 	value &= ~RP_ECTL_2_R2_RX_CTLE_1C_MASK;
 	value |= soc->ectl.regs.rp_ectl_2_r2;
@@ -615,6 +917,26 @@ static void tegra_pcie_program_ectl_sett
 	writel(value, port->base + RP_ECTL_6_R2);
 }
 
+static void tegra_pcie_enable_wrap(void)
+{
+	u32 val;
+	void __iomem *msel_base;
+
+#define MSELECT_CONFIG_BASE     0x50060000
+#define MSELECT_CONFIG_WRAP_TO_INCR_SLAVE1      BIT(28)
+#define MSELECT_CONFIG_ERR_RESP_EN_SLAVE1       BIT(24)
+
+	/* Config MSELECT to support wrap trans for normal NC & GRE mapping */
+	msel_base = ioremap(MSELECT_CONFIG_BASE, 4);
+	val = readl(msel_base);
+	/* Enable WRAP_TO_INCR_SLAVE1 */
+	val |= MSELECT_CONFIG_WRAP_TO_INCR_SLAVE1;
+	/* Disable ERR_RESP_EN_SLAVE1 */
+	val &= ~MSELECT_CONFIG_ERR_RESP_EN_SLAVE1;
+	writel(val, msel_base);
+	iounmap(msel_base);
+}
+
 static void tegra_pcie_apply_sw_fixup(struct tegra_pcie_port *port)
 {
 	const struct tegra_pcie_soc *soc = port->pcie->soc;
@@ -649,6 +971,76 @@ static void tegra_pcie_apply_sw_fixup(st
 	value &= ~PCI_EXP_LNKSTA_CLS;
 	value |= PCI_EXP_LNKSTA_CLS_2_5GB;
 	writel(value, port->base + RP_LINK_CONTROL_STATUS_2);
+
+	if (soc->enable_wrap)
+		tegra_pcie_enable_wrap();
+
+	if (soc->has_aspm_l1ss) {
+		/* Set port Common_Mode_Restore_Time to 30us */
+		value = readl(port->base + RP_L1_PM_SUBSTATES_CTL);
+		value &= ~RP_L1_PM_SUBSTATES_CTL_CM_RTIME_MASK;
+		value |= (0x1E << RP_L1_PM_SUBSTATES_CTL_CM_RTIME_SHIFT);
+		writel(value, port->base + RP_L1_PM_SUBSTATES_CTL);
+
+		/* set port T_POWER_ON to 70us */
+		value = readl(port->base + RP_L1_PM_SUBSTATES_CTL);
+		value &= ~(RP_L1_PM_SUBSTATES_CTL_T_PWRN_SCL_MASK |
+				RP_L1_PM_SUBSTATES_CTL_T_PWRN_VAL_MASK);
+		value |= (1 << RP_L1_PM_SUBSTATES_CTL_T_PWRN_SCL_SHIFT) |
+			(7 << RP_L1_PM_SUBSTATES_CTL_T_PWRN_VAL_SHIFT);
+		writel(value, port->base + RP_L1_PM_SUBSTATES_CTL);
+
+		/* Following is based on clk_m being 19.2 MHz */
+		value = readl(port->base + RP_TIMEOUT0);
+		value &= ~RP_TIMEOUT0_PAD_PWRUP_MASK;
+		value |= RP_TIMEOUT0_PAD_PWRUP;
+		value &= ~RP_TIMEOUT0_PAD_PWRUP_CM_MASK;
+		value |= RP_TIMEOUT0_PAD_PWRUP_CM;
+		value &= ~RP_TIMEOUT0_PAD_SPDCHNG_GEN2_MASK;
+		value |= RP_TIMEOUT0_PAD_SPDCHNG_GEN2;
+		writel(value, port->base + RP_TIMEOUT0);
+
+		value = readl(port->base + RP_TIMEOUT1);
+		value &= ~RP_TIMEOUT1_RCVRY_SPD_SUCCESS_EIDLE_MASK;
+		value |= RP_TIMEOUT1_RCVRY_SPD_SUCCESS_EIDLE;
+		value &= ~RP_TIMEOUT1_RCVRY_SPD_UNSUCCESS_EIDLE_MASK;
+		value |= RP_TIMEOUT1_RCVRY_SPD_UNSUCCESS_EIDLE;
+		writel(value, port->base + RP_TIMEOUT1);
+
+		value = readl(port->base + RP_XP_REF);
+		value &= ~RP_XP_REF_MICROSECOND_LIMIT_MASK;
+		value |= RP_XP_REF_MICROSECOND_LIMIT;
+		value |= RP_XP_REF_MICROSECOND_ENABLE;
+		value |= RP_XP_REF_CPL_TO_OVERRIDE;
+		value &= ~RP_XP_REF_CPL_TO_CUSTOM_VALUE_MASK;
+		value |= RP_XP_REF_CPL_TO_CUSTOM_VALUE;
+		writel(value, port->base + RP_XP_REF);
+
+		value = readl(port->base + RP_L1_PM_SUBSTATES_1_CTL);
+		value &= ~RP_L1_PM_SUBSTATES_1_CTL_PWR_OFF_DLY_MASK;
+		value |= RP_L1_PM_SUBSTATES_1_CTL_PWR_OFF_DLY;
+		writel(value, port->base + RP_L1_PM_SUBSTATES_1_CTL);
+
+		value = readl(port->base + RP_L1_PM_SUBSTATES_2_CTL);
+		value &= ~RP_L1_PM_SUBSTATES_2_CTL_T_L1_2_DLY_MASK;
+		value |= RP_L1_PM_SUBSTATES_2_CTL_T_L1_2_DLY;
+		value &= ~RP_L1_PM_SUBSTATES_2_CTL_MICROSECOND_MASK;
+		value |= RP_L1_PM_SUBSTATES_2_CTL_MICROSECOND;
+		value &= ~RP_L1_PM_SUBSTATES_2_CTL_MICROSECOND_COMP_MASK;
+		value |= RP_L1_PM_SUBSTATES_2_CTL_MICROSECOND_COMP;
+		writel(value, port->base + RP_L1_PM_SUBSTATES_2_CTL);
+	}
+
+	if (soc->l1ss_rp_wake_fixup) {
+		/*
+		 * Set CLKREQ asserted delay greater than Power_Off
+		 * time (2us) to avoid RP wakeup in L1.2.ENTRY
+		 */
+		value = readl(port->base + RP_L1_PM_SUBSTATES_1_CTL);
+		value &= ~RP_L1SS_1_CTL_CLKREQ_ASSERTED_DLY_MASK;
+		value |= RP_L1SS_1_CTL_CLKREQ_ASSERTED_DLY;
+		writel(value, port->base + RP_L1_PM_SUBSTATES_1_CTL);
+	}
 }
 
 static void tegra_pcie_port_enable(struct tegra_pcie_port *port)
@@ -661,8 +1053,12 @@ static void tegra_pcie_port_enable(struc
 	value = afi_readl(port->pcie, ctrl);
 	value |= AFI_PEX_CTRL_REFCLK_EN;
 
-	if (soc->has_pex_clkreq_en)
-		value |= AFI_PEX_CTRL_CLKREQ_EN;
+	if (soc->has_pex_clkreq_en) {
+		if (port->supports_clkreq)
+			value &= ~AFI_PEX_CTRL_CLKREQ_EN;
+		else
+			value |= AFI_PEX_CTRL_CLKREQ_EN;
+	}
 
 	value |= AFI_PEX_CTRL_OVERRIDE_EN;
 
@@ -670,6 +1066,14 @@ static void tegra_pcie_port_enable(struc
 
 	tegra_pcie_port_reset(port);
 
+	/*
+	 * On platforms where MXM is not directly connected to Tegra root port,
+	 * 200 ms delay (worst case) is required after reset, to ensure linkup
+	 * between PCIe switch and MXM
+	 */
+	if (port->has_mxm_port)
+		mdelay(200);
+
 	if (soc->force_pca_enable) {
 		value = readl(port->base + RP_VEND_CTL2);
 		value |= RP_VEND_CTL2_PCA_ENABLE;
@@ -1092,10 +1496,20 @@ static void tegra_pcie_enable_controller
 	unsigned long value;
 
 	/* enable PLL power down */
-	if (pcie->phy) {
+	if (soc->has_aspm_l1ss) {
 		value = afi_readl(pcie, AFI_PLLE_CONTROL);
 		value &= ~AFI_PLLE_CONTROL_BYPASS_PADS2PLLE_CONTROL;
 		value |= AFI_PLLE_CONTROL_PADS2PLLE_CONTROL_EN;
+
+		list_for_each_entry(port, &pcie->ports, list) {
+			if (!port->supports_clkreq) {
+				value &= ~AFI_PLLE_CONTROL_PADS2PLLE_CONTROL_EN;
+				break;
+			}
+		}
+
+		value &= ~AFI_PLLE_CONTROL_BYPASS_PCIE2PLLE_CONTROL;
+		value |= AFI_PLLE_CONTROL_PCIE2PLLE_CONTROL_EN;
 		afi_writel(pcie, value, AFI_PLLE_CONTROL);
 	}
 
@@ -1506,6 +1920,17 @@ static int tegra_pcie_put_resources(stru
 	return 0;
 }
 
+static void tegra_pcie_config_plat(struct tegra_pcie *pcie, bool set)
+{
+	struct tegra_pcie_port *port;
+	int count;
+
+	list_for_each_entry(port, &pcie->ports, list) {
+		for (count = 0; count < port->n_gpios; ++count)
+			gpiod_set_value(gpio_to_desc(port->gpios[count]), set);
+	}
+}
+
 static void tegra_pcie_pme_turnoff(struct tegra_pcie_port *port)
 {
 	struct tegra_pcie *pcie = port->pcie;
@@ -1530,6 +1955,11 @@ static void tegra_pcie_pme_turnoff(struc
 	val = afi_readl(pcie, AFI_PCIE_PME);
 	val &= ~(0x1 << soc->ports[port->index].pme.turnoff_bit);
 	afi_writel(pcie, val, AFI_PCIE_PME);
+
+	/* PCIe link is in L2, bypass CLKREQ# control over PLLE power down */
+	val = afi_readl(pcie, AFI_PLLE_CONTROL);
+	val |= AFI_PLLE_CONTROL_BYPASS_PADS2PLLE_CONTROL;
+	afi_writel(pcie, val, AFI_PLLE_CONTROL);
 }
 
 static void tegra_pcie_msi_irq(struct irq_desc *desc)
@@ -1885,6 +2315,10 @@ static int tegra_pcie_get_xbar_config(st
 			*xbar = AFI_PCIE_CONFIG_SM2TMS0_XBAR_CONFIG_211;
 			return 0;
 		}
+	} else if (of_device_is_compatible(np, "nvidia,tegra210b01-pcie")) {
+		dev_info(dev, "4x1, 1x1 configuration\n");
+		*xbar = AFI_PCIE_CONFIG_SM2TMS0_XBAR_CONFIG_X4_X1;
+		return 0;
 	} else if (of_device_is_compatible(np, "nvidia,tegra124-pcie") ||
 		   of_device_is_compatible(np, "nvidia,tegra210-pcie")) {
 		switch (lanes) {
@@ -2121,13 +2555,36 @@ static int tegra_pcie_parse_dt(struct te
 	unsigned int lane = 0;
 	int err;
 
+	pcie->pex_wake = of_get_named_gpio(np, "nvidia,wake-gpio", 0);
+	if (gpio_is_valid(pcie->pex_wake)) {
+		err = devm_gpio_request(dev, pcie->pex_wake, "pex_wake");
+		if (err < 0) {
+			dev_err(dev, "pex_wake gpio request failed: %d\n", err);
+			return err;
+		}
+		err = gpio_direction_input(pcie->pex_wake);
+		if (err < 0) {
+			dev_err(dev, "pex_wake set gpio dir input failed: %d\n",
+				err);
+			return err;
+		}
+	}
+
 	/* parse root ports */
 	for_each_child_of_node(np, port) {
 		struct tegra_pcie_port *rp;
 		unsigned int index;
+		const char *type;
 		u32 value;
 		char *label;
 
+		if (!of_property_read_string(port, "device_type", &type)) {
+			if (strncmp(type, "pci", sizeof("pci")))
+				continue;
+		} else {
+			continue;
+		}
+
 		err = of_pci_get_devfn(port);
 		if (err < 0) {
 			dev_err(dev, "failed to parse address: %d\n", err);
@@ -2217,6 +2674,68 @@ static int tegra_pcie_parse_dt(struct te
 			}
 		}
 
+		rp->n_gpios = of_gpio_named_count(port, "nvidia,plat-gpios");
+		if (rp->n_gpios > 0) {
+			int count, gpio;
+			enum of_gpio_flags flags;
+			unsigned long f;
+
+			rp->gpios = devm_kzalloc(dev, rp->n_gpios * sizeof(int),
+						 GFP_KERNEL);
+			if (!rp->gpios)
+				return -ENOMEM;
+
+			for (count = 0; count < rp->n_gpios; ++count) {
+				gpio = of_get_named_gpio_flags(port,
+							"nvidia,plat-gpios",
+							count, &flags);
+				if (!gpio_is_valid(gpio))
+					return gpio;
+
+				f = (flags & OF_GPIO_ACTIVE_LOW) ?
+				    (GPIOF_OUT_INIT_LOW | GPIOF_ACTIVE_LOW) :
+				    GPIOF_OUT_INIT_HIGH;
+
+				err = devm_gpio_request_one(dev, gpio, f, NULL);
+				if (err < 0) {
+					dev_err(dev, "gpio %d request failed\n",
+						gpio);
+					return err;
+				}
+				rp->gpios[count] = gpio;
+			}
+		}
+
+		rp->has_mxm_port = of_property_read_bool(port,
+							 "nvidia,has-mxm-port");
+		if (rp->has_mxm_port) {
+			rp->pwr_gd_gpio = of_get_named_gpio(port,
+						"nvidia,pwr-gd-gpio", 0);
+			if (gpio_is_valid(rp->pwr_gd_gpio)) {
+				err = devm_gpio_request(dev, rp->pwr_gd_gpio,
+							"pwr_gd_gpio");
+				if (err < 0) {
+					dev_err(dev, "%s: pwr_gd_gpio request failed %d\n",
+						__func__, err);
+					return err;
+				}
+
+				err = gpio_direction_input(rp->pwr_gd_gpio);
+				if (err < 0) {
+					dev_err(dev, "%s: pwr_gd_gpio direction input failed %d\n",
+						__func__, err);
+				}
+			}
+		}
+
+		err = of_property_read_u32(port, "nvidia,disable-aspm-states",
+					   &rp->aspm_state);
+		if (err < 0)
+			rp->aspm_state = 0;
+
+		rp->supports_clkreq = of_property_read_bool(port,
+							    "supports-clkreq");
+
 		list_add_tail(&rp->list, &pcie->ports);
 	}
 
@@ -2348,6 +2867,86 @@ static void tegra_pcie_change_link_speed
 	}
 }
 
+static int tegra_pcie_scale_freq(struct tegra_pcie *pcie)
+{
+	struct tegra_pcie_port *port, *tmp;
+	const struct tegra_pcie_soc *soc = pcie->soc;
+	int err = 0;
+	u32 val = 0;
+	u32 active_lanes = 0;
+	bool is_gen2 = false;
+
+	list_for_each_entry_safe(port, tmp, &pcie->ports, list) {
+		val = readl(port->base + RP_LINK_CONTROL_STATUS);
+		active_lanes += ((val &
+					RP_LINK_CONTROL_STATUS_NEG_LINK_WIDTH) >> 20);
+		if (((val & RP_LINK_CONTROL_STATUS_LINK_SPEED) >> 16) == 2)
+			is_gen2 = true;
+	}
+
+	if (soc->dvfs_mselect) {
+		struct clk *mselect_clk;
+
+		active_lanes = 0;
+		dev_dbg(pcie->dev, "mselect_clk is set @ %u\n",
+				soc->dfs_tbl[active_lanes][is_gen2].afi_clk);
+		mselect_clk = devm_clk_get(pcie->dev, "mselect");
+		if (IS_ERR(mselect_clk)) {
+			dev_err(pcie->dev, "mselect clk_get failed: %ld\n",
+					PTR_ERR(mselect_clk));
+			return PTR_ERR(mselect_clk);
+		}
+		err = clk_set_rate(mselect_clk,
+				soc->dfs_tbl[active_lanes][is_gen2].afi_clk);
+		if (err) {
+			dev_err(pcie->dev,
+					"setting mselect clk to %u failed : %d\n",
+					soc->dfs_tbl[active_lanes][is_gen2].afi_clk,
+					err);
+			return err;
+		}
+	}
+
+	if (soc->dvfs_afi) {
+		dev_dbg(pcie->dev, "afi_clk is set @ %u\n",
+				soc->dfs_tbl[active_lanes][is_gen2].afi_clk);
+		err = clk_set_rate(devm_clk_get(pcie->dev, "afi"),
+				soc->dfs_tbl[active_lanes][is_gen2].afi_clk);
+		if (err) {
+			dev_err(pcie->dev,
+					"setting afi clk to %u failed : %d\n",
+					soc->dfs_tbl[active_lanes][is_gen2].afi_clk,
+					err);
+			return err;
+		}
+	}
+
+	dev_dbg(pcie->dev, "emc_clk is set @ %u\n",
+			soc->dfs_tbl[active_lanes][is_gen2].emc_clk);
+#if IS_ENABLED(CONFIG_TEGRA_BWMGR)
+	err = tegra_bwmgr_set_emc(pcie->emc_bwmgr,
+			soc->dfs_tbl[active_lanes][is_gen2].emc_clk,
+			TEGRA_BWMGR_SET_EMC_FLOOR);
+	if (err < 0) {
+		dev_err(pcie->dev, "setting emc clk to %u failed : %d\n",
+				soc->dfs_tbl[active_lanes][is_gen2].emc_clk, err);
+		return err;
+	}
+#endif
+
+	return err;
+}
+
+static int tegra_pcie_mxm_pwr_init(struct tegra_pcie_port *port)
+{
+	mdelay(100);
+
+	if (!gpio_get_value(port->pwr_gd_gpio))
+		return 1;
+
+	return 0;
+}
+
 static void tegra_pcie_enable_ports(struct tegra_pcie *pcie)
 {
 	struct device *dev = pcie->dev;
@@ -2375,6 +2974,8 @@ static void tegra_pcie_enable_ports(stru
 
 	if (pcie->soc->has_gen2)
 		tegra_pcie_change_link_speed(pcie);
+
+	tegra_pcie_scale_freq(pcie);
 }
 
 static void tegra_pcie_disable_ports(struct tegra_pcie *pcie)
@@ -2410,6 +3011,12 @@ static const struct tegra_pcie_soc tegra
 	.program_deskew_time = false,
 	.update_fc_timer = false,
 	.has_cache_bars = true,
+	.enable_wrap = false,
+	.has_aspm_l1 = false,
+	.has_aspm_l1ss = false,
+	.l1ss_rp_wake_fixup = false,
+	.dvfs_mselect = false,
+	.dvfs_afi = false,
 	.ectl.enable = false,
 };
 
@@ -2439,6 +3046,12 @@ static const struct tegra_pcie_soc tegra
 	.program_deskew_time = false,
 	.update_fc_timer = false,
 	.has_cache_bars = false,
+	.enable_wrap = false,
+	.has_aspm_l1 = true,
+	.has_aspm_l1ss = false,
+	.l1ss_rp_wake_fixup = false,
+	.dvfs_mselect = false,
+	.dvfs_afi = false,
 	.ectl.enable = false,
 };
 
@@ -2460,6 +3073,12 @@ static const struct tegra_pcie_soc tegra
 	.program_deskew_time = false,
 	.update_fc_timer = false,
 	.has_cache_bars = false,
+	.enable_wrap = false,
+	.has_aspm_l1 = true,
+	.has_aspm_l1ss = false,
+	.l1ss_rp_wake_fixup = false,
+	.dvfs_mselect = false,
+	.dvfs_afi = false,
 	.ectl.enable = false,
 };
 
@@ -2483,12 +3102,22 @@ static const struct tegra_pcie_soc tegra
 	.program_deskew_time = true,
 	.update_fc_timer = true,
 	.has_cache_bars = false,
+	.enable_wrap = true,
+	.has_aspm_l1 = true,
+	.has_aspm_l1ss = true,
+	.l1ss_rp_wake_fixup = true,
+	.dvfs_mselect = true,
+	.dvfs_afi = false,
+	.dfs_tbl = {
+		{{204000000, 102000000}, {408000000, 528000000} } },
 	.ectl = {
 		.regs = {
+			.rp_ectl_1_r1 = 0x0000001f,
 			.rp_ectl_2_r1 = 0x0000000f,
 			.rp_ectl_4_r1 = 0x00000067,
 			.rp_ectl_5_r1 = 0x55010000,
 			.rp_ectl_6_r1 = 0x00000001,
+			.rp_ectl_1_r2 = 0x0000001f,
 			.rp_ectl_2_r2 = 0x0000008f,
 			.rp_ectl_4_r2 = 0x000000c7,
 			.rp_ectl_5_r2 = 0x55010000,
@@ -2498,6 +3127,51 @@ static const struct tegra_pcie_soc tegra
 	},
 };
 
+static const struct tegra_pcie_soc tegra210b01_pcie = {
+	.num_ports = 2,
+	.ports = tegra20_pcie_ports,
+	.msi_base_shift = 8,
+	.pads_pll_ctl = PADS_PLL_CTL_TEGRA30,
+	.tx_ref_sel = PADS_PLL_CTL_TXCLKREF_BUF_EN,
+	.pads_refclk_cfg0 = 0x90b890b8,
+	/* FC threshold is bit[25:18] */
+	.update_fc_threshold = 0x01800000,
+	.has_pex_clkreq_en = true,
+	.has_pex_bias_ctrl = true,
+	.has_intr_prsnt_sense = true,
+	.has_cml_clk = true,
+	.has_gen2 = true,
+	.force_pca_enable = true,
+	.program_uphy = true,
+	.update_clamp_threshold = false,
+	.program_deskew_time = true,
+	.update_fc_timer = true,
+	.has_cache_bars = false,
+	.enable_wrap = false,
+	.has_aspm_l1 = true,
+	.has_aspm_l1ss = true,
+	.l1ss_rp_wake_fixup = true,
+	.dvfs_mselect = true,
+	.dvfs_afi = false,
+	.dfs_tbl = {
+		{{204000000, 102000000}, {408000000, 528000000} } },
+	.ectl = {
+		.regs = {
+			.rp_ectl_1_r1 = 0x00000027,
+			.rp_ectl_2_r1 = 0x0000000f,
+			.rp_ectl_4_r1 = 0x00000067,
+			.rp_ectl_5_r1 = 0x00000000,
+			.rp_ectl_6_r1 = 0x00000000,
+			.rp_ectl_1_r2 = 0x00000027,
+			.rp_ectl_2_r2 = 0x0000008f,
+			.rp_ectl_4_r2 = 0x000000c7,
+			.rp_ectl_5_r2 = 0x00000000,
+			.rp_ectl_6_r2 = 0x00000000,
+		},
+		.enable = true,
+	},
+};
+
 static const struct tegra_pcie_port_soc tegra186_pcie_ports[] = {
 	{ .pme.turnoff_bit =  0, .pme.ack_bit =  5 },
 	{ .pme.turnoff_bit =  8, .pme.ack_bit = 10 },
@@ -2524,11 +3198,25 @@ static const struct tegra_pcie_soc tegra
 	.program_deskew_time = false,
 	.update_fc_timer = false,
 	.has_cache_bars = false,
+	.enable_wrap = false,
+	.has_aspm_l1 = true,
+	.has_aspm_l1ss = true,
+	.l1ss_rp_wake_fixup = false,
+	.dvfs_mselect = false,
+	.dvfs_afi = true,
+	.dfs_tbl = {
+		{{0, 0}, {0, 0} },
+		{{102000000, 480000000}, {102000000, 480000000} },
+		{{102000000, 480000000}, {204000000, 480000000} },
+		{{102000000, 480000000}, {204000000, 480000000} },
+		{{204000000, 480000000}, {408000000, 480000000} },
+		{{204000000, 480000000}, {408000000, 640000000} } },
 	.ectl.enable = false,
 };
 
 static const struct of_device_id tegra_pcie_of_match[] = {
 	{ .compatible = "nvidia,tegra186-pcie", .data = &tegra186_pcie },
+	{ .compatible = "nvidia,tegra210b01-pcie", .data = &tegra210b01_pcie },
 	{ .compatible = "nvidia,tegra210-pcie", .data = &tegra210_pcie },
 	{ .compatible = "nvidia,tegra124-pcie", .data = &tegra124_pcie },
 	{ .compatible = "nvidia,tegra30-pcie", .data = &tegra30_pcie },
@@ -2537,6 +3225,937 @@ static const struct of_device_id tegra_p
 };
 MODULE_DEVICE_TABLE(of, tegra_pcie_of_match);
 
+static int list_devices(struct seq_file *s, void *data)
+{
+	struct pci_dev *pdev = NULL;
+	u16 vendor, device, devclass, speed;
+	bool pass = false;
+	int ret = 0;
+
+	for_each_pci_dev(pdev) {
+		pass = true;
+		ret = pci_read_config_word(pdev, PCI_VENDOR_ID, &vendor);
+		if (ret) {
+			pass = false;
+			break;
+		}
+		ret = pci_read_config_word(pdev, PCI_DEVICE_ID, &device);
+		if (ret) {
+			pass = false;
+			break;
+		}
+		ret = pci_read_config_word(pdev, PCI_CLASS_DEVICE, &devclass);
+		if (ret) {
+			pass = false;
+			break;
+		}
+		pcie_capability_read_word(pdev, PCI_EXP_LNKSTA, &speed);
+
+		seq_printf(s, "%s  Vendor:%04x  Device id:%04x  ",
+				kobject_name(&pdev->dev.kobj), vendor,
+				device);
+		seq_printf(s, "Class:%04x  Speed:%s  Driver:%s(%s)\n", devclass,
+				((speed & PCI_EXP_LNKSTA_CLS_5_0GB) ==
+				 PCI_EXP_LNKSTA_CLS_5_0GB) ?
+				"Gen2" : "Gen1",
+				(pdev->driver) ? "enabled" : "disabled",
+				(pdev->driver) ? pdev->driver->name : NULL);
+	}
+	if (!pass)
+		seq_puts(s, "Couldn't read devices\n");
+
+	return ret;
+}
+
+static void tegra_pcie_link_speed(struct tegra_pcie *pcie, bool is_gen2)
+{
+	struct device *dev = pcie->dev;
+	struct tegra_pcie_port *port, *tmp;
+	ktime_t deadline;
+	unsigned long val;
+
+	list_for_each_entry_safe(port, tmp, &pcie->ports, list) {
+		/*
+		 * Link Capabilities 2 register is hardwired to 0 in Tegra,
+		 * so no need to read it before setting target speed.
+		 */
+		val = readl(port->base + RP_LINK_CONTROL_STATUS_2);
+		val &= ~PCI_EXP_LNKSTA_CLS;
+		if (is_gen2)
+			val |= PCI_EXP_LNKSTA_CLS_5_0GB;
+		else
+			val |= PCI_EXP_LNKSTA_CLS_2_5GB;
+		writel(val, port->base + RP_LINK_CONTROL_STATUS_2);
+
+		/*
+		 * Poll until link comes back from recovery to avoid race
+		 * condition.
+		 */
+		deadline = ktime_add_us(ktime_get(), LINK_RETRAIN_TIMEOUT);
+		for (;;) {
+			val = readl(port->base + RP_LINK_CONTROL_STATUS);
+			if (!(val & PCI_EXP_LNKSTA_LT))
+				break;
+			if (ktime_after(ktime_get(), deadline))
+				break;
+			usleep_range(2000, 3000);
+		}
+		if (val & PCI_EXP_LNKSTA_LT)
+			dev_err(dev, "PCIe port %u link is still in recovery\n",
+					port->index);
+
+		/* Retrain the link */
+		val = readl(port->base + RP_LINK_CONTROL_STATUS);
+		val |= PCI_EXP_LNKCTL_RL;
+		writel(val, port->base + RP_LINK_CONTROL_STATUS);
+
+		deadline = ktime_add_us(ktime_get(), LINK_RETRAIN_TIMEOUT);
+		for (;;) {
+			val = readl(port->base + RP_LINK_CONTROL_STATUS);
+			if (!(val & PCI_EXP_LNKSTA_LT))
+				break;
+			if (ktime_after(ktime_get(), deadline))
+				break;
+			usleep_range(2000, 3000);
+		}
+		if (val & PCI_EXP_LNKSTA_LT)
+			dev_err(dev, "link retrain of PCIe port %u failed\n",
+					port->index);
+	}
+}
+
+static int apply_link_speed(struct seq_file *s, void *data)
+{
+	struct tegra_pcie *pcie = (struct tegra_pcie *)(s->private);
+
+	seq_printf(s, "Changing link speed to %s... ",
+			(is_gen2_speed) ? "Gen2" : "Gen1");
+	tegra_pcie_link_speed(pcie, is_gen2_speed);
+	seq_puts(s, "Done\n");
+	return 0;
+}
+
+static int check_d3hot(struct seq_file *s, void *data)
+{
+	u16 val;
+	struct pci_dev *pdev = NULL;
+
+	/* Force all the devices (including RPs) in d3 hot state */
+	for_each_pci_dev(pdev) {
+		if (pci_pcie_type(pdev) == PCI_EXP_TYPE_ROOT_PORT ||
+				pci_pcie_type(pdev) == PCI_EXP_TYPE_DOWNSTREAM)
+			continue;
+		/* First, keep Downstream component in D3_Hot */
+		pci_read_config_word(pdev, pdev->pm_cap + PCI_PM_CTRL,
+				&val);
+		if ((val & PCI_PM_CTRL_STATE_MASK) == (__force u16)PCI_D3hot)
+			seq_printf(s, "device[%x:%x] is already in D3_hot]\n",
+					pdev->vendor, pdev->device);
+		val &= ~PCI_PM_CTRL_STATE_MASK;
+		val |= (__force u16)PCI_D3hot;
+		pci_write_config_word(pdev, pdev->pm_cap + PCI_PM_CTRL, val);
+		/* Keep corresponding upstream component in D3_Hot */
+		pci_read_config_word(pdev->bus->self,
+				pdev->bus->self->pm_cap + PCI_PM_CTRL,
+				&val);
+		val &= ~PCI_PM_CTRL_STATE_MASK;
+		val |= (__force u16)PCI_D3hot;
+		pci_write_config_word(pdev->bus->self,
+				pdev->bus->self->pm_cap + PCI_PM_CTRL,
+				val);
+		mdelay(100);
+		/* check if they have changed their state */
+		pci_read_config_word(pdev, pdev->pm_cap + PCI_PM_CTRL, &val);
+		if ((val & PCI_PM_CTRL_STATE_MASK) == (__force u16)PCI_D3hot)
+			seq_printf(s, "device[%x:%x] transitioned to D3_hot]\n",
+					pdev->vendor, pdev->device);
+		else
+			seq_printf(s, "device[%x:%x] couldn't transition to D3_hot]\n",
+					pdev->vendor, pdev->device);
+		pci_read_config_word(pdev->bus->self,
+				pdev->bus->self->pm_cap + PCI_PM_CTRL,
+				&val);
+		if ((val & PCI_PM_CTRL_STATE_MASK) == (__force u16)PCI_D3hot)
+			seq_printf(s, "device[%x:%x] transitioned to D3_hot]\n",
+					pdev->bus->self->vendor,
+					pdev->bus->self->device);
+		else
+			seq_printf(s, "device[%x:%x] couldn't transition to D3_hot]\n",
+					pdev->bus->self->vendor,
+					pdev->bus->self->device);
+	}
+
+	return 0;
+}
+
+static int dump_config_space(struct seq_file *s, void *data)
+{
+	u8 val;
+	int row, col;
+	struct pci_dev *pdev = NULL;
+
+	for_each_pci_dev(pdev) {
+		int row_cnt = pci_is_pcie(pdev) ?
+			PCI_EXT_CFG_SPACE_SIZE : PCI_CFG_SPACE_SIZE;
+		seq_printf(s, "%s\n", kobject_name(&pdev->dev.kobj));
+		seq_printf(s, "%s\n", "------------");
+
+		for (row = 0; row < (row_cnt / 16); row++) {
+			seq_printf(s, "%02x: ", (row * 16));
+			for (col = 0; col < 16; col++) {
+				pci_read_config_byte(pdev, ((row * 16) + col),
+						&val);
+				seq_printf(s, "%02x ", val);
+			}
+			seq_puts(s, "\n");
+		}
+	}
+	return 0;
+}
+
+static int dump_afi_space(struct seq_file *s, void *data)
+{
+	u32 val, offset;
+	struct tegra_pcie_port *port = NULL;
+	struct tegra_pcie *pcie = (struct tegra_pcie *)(s->private);
+
+	list_for_each_entry(port, &pcie->ports, list) {
+		seq_puts(s, "Offset:  Values\n");
+		for (offset = 0; offset < 0x200; offset += 0x10) {
+			val = afi_readl(port->pcie, offset);
+			seq_printf(s, "%6x: %8x %8x %8x %8x\n", offset,
+					afi_readl(port->pcie, offset),
+					afi_readl(port->pcie, offset + 4),
+					afi_readl(port->pcie, offset + 8),
+					afi_readl(port->pcie, offset + 12));
+		}
+	}
+	return 0;
+}
+
+static int config_read(struct seq_file *s, void *data)
+{
+	struct pci_dev *pdev = NULL;
+
+	pdev = pci_get_domain_bus_and_slot(0, (bdf >> 8), (bdf & 0xFF));
+	if (!pdev) {
+		seq_printf(s, "%02d:%02d.%02d : Doesn't exist\n",
+				(bdf >> 8), PCI_SLOT(bdf), PCI_FUNC(bdf));
+		seq_puts(s, "Enter (bus<<8 | dev<<3 | fn) val to bdf file\n");
+		goto end;
+	}
+	if (config_offset >= PCI_EXT_CFG_SPACE_SIZE) {
+		seq_printf(s, "Config offset exceeds max (i.e %d) value\n",
+				PCI_EXT_CFG_SPACE_SIZE);
+	}
+	if (!(config_offset & 0x3)) {
+		u32 val;
+		/* read 32 */
+		pci_read_config_dword(pdev, config_offset, &val);
+		seq_printf(s, "%08x\n", val);
+		config_val = val;
+	} else if (!(config_offset & 0x1)) {
+		u16 val;
+		/* read 16 */
+		pci_read_config_word(pdev, config_offset, &val);
+		seq_printf(s, "%04x\n", val);
+		config_val = val;
+	} else {
+		u8 val;
+		/* read 8 */
+		pci_read_config_byte(pdev, config_offset, &val);
+		seq_printf(s, "%02x\n", val);
+		config_val = val;
+	}
+
+end:
+	return 0;
+}
+
+static int config_write(struct seq_file *s, void *data)
+{
+	struct pci_dev *pdev = NULL;
+
+	pdev = pci_get_domain_bus_and_slot(0, (bdf >> 8), (bdf & 0xFF));
+	if (!pdev) {
+		seq_printf(s, "%02d:%02d.%02d : Doesn't exist\n",
+				(bdf >> 8), PCI_SLOT(bdf), PCI_FUNC(bdf));
+		seq_puts(s, "Enter (bus<<8 | dev<<3 | fn) val to bdf file\n");
+		goto end;
+	}
+	if (config_offset >= PCI_EXT_CFG_SPACE_SIZE) {
+		seq_printf(s, "Config offset exceeds max (i.e %d) value\n",
+				PCI_EXT_CFG_SPACE_SIZE);
+	}
+	if (!(config_offset & 0x3)) {
+		/* write 32 */
+		pci_write_config_dword(pdev, config_offset, config_val);
+	} else if (!(config_offset & 0x1)) {
+		/* write 16 */
+		pci_write_config_word(pdev, config_offset,
+				(u16)(config_val & 0xFFFF));
+	} else {
+		/* write 8 */
+		pci_write_config_byte(pdev, config_offset,
+				(u8)(config_val & 0xFF));
+	}
+
+end:
+	return 0;
+}
+
+static int power_down(struct seq_file *s, void *data)
+{
+	struct tegra_pcie_port *port = NULL;
+	struct tegra_pcie *pcie = (struct tegra_pcie *)(s->private);
+	const struct tegra_pcie_soc *soc = pcie->soc;
+	bool pass = true;
+	u8 ack_bit;
+	u32 val;
+	int err;
+
+	list_for_each_entry(port, &pcie->ports, list) {
+		val = afi_readl(pcie, AFI_PCIE_PME);
+		val |= (0x1 << soc->ports[port->index].pme.turnoff_bit);
+		afi_writel(pcie, val, AFI_PCIE_PME);
+
+		ack_bit = soc->ports[port->index].pme.ack_bit;
+		err = readl_poll_timeout(pcie->afi + AFI_PCIE_PME, val,
+					 val & (0x1 << ack_bit), 1,
+					 PME_ACK_TIMEOUT);
+		if (err)
+			dev_err(pcie->dev, "PME Ack is not received on port: %d\n",
+				port->index);
+
+		usleep_range(10000, 11000);
+
+		val = afi_readl(pcie, AFI_PCIE_PME);
+		val &= ~(0x1 << soc->ports[port->index].pme.turnoff_bit);
+		afi_writel(pcie, val, AFI_PCIE_PME);
+
+		mdelay(1000);
+
+		val = readl(port->base + RP_LTSSM_DBGREG);
+		if (!(val & RP_LTSSM_DBGREG_LINKFSM16)) {
+			pass = false;
+			goto out;
+		}
+	}
+
+out:
+	if (pass)
+		seq_puts(s, "[pass: pcie_power_down]\n");
+	else
+		seq_puts(s, "[fail: pcie_power_down]\n");
+	pr_info("PCIE power down test END..\n");
+	return 0;
+}
+
+static int loopback(struct seq_file *s, void *data)
+{
+	struct tegra_pcie_port *port = (struct tegra_pcie_port *)(s->private);
+	unsigned int new, i, val;
+
+	new = readl(port->base + RP_LINK_CONTROL_STATUS);
+
+	if (!(new & RP_LINK_CONTROL_STATUS_DL_LINK_ACTIVE)) {
+		pr_info("PCIE port %d not active\n", port->index);
+		return -EINVAL;
+	}
+
+	/* trigger trace ram on loopback states */
+	val = LTSSM_TRACE_CONTROL_CLEAR_STORE_EN |
+		LTSSM_TRACE_CONTROL_TRIG_ON_EVENT |
+		(0x08 << LTSSM_TRACE_CONTROL_TRIG_LTSSM_MAJOR_OFFSET) |
+		(0x00 << LTSSM_TRACE_CONTROL_TRIG_PTX_LTSSM_MINOR_OFFSET) |
+		(0x00 << LTSSM_TRACE_CONTROL_TRIG_PRX_LTSSM_MAJOR_OFFSET);
+	writel(val, port->base + RP_LTSSM_TRACE_CONTROL);
+
+	/* clear trace ram */
+	val = readl(port->base + RP_LTSSM_TRACE_CONTROL);
+	val |= LTSSM_TRACE_CONTROL_CLEAR_RAM;
+	writel(val, port->base + RP_LTSSM_TRACE_CONTROL);
+	val &= ~LTSSM_TRACE_CONTROL_CLEAR_RAM;
+	writel(val, port->base + RP_LTSSM_TRACE_CONTROL);
+
+	/* reset and clear status */
+	port->loopback_stat = 0;
+
+	new = readl(port->base + RP_VEND_XP);
+	new &= ~RP_VEND_XP_PRBS_EN;
+	writel(new, port->base + RP_VEND_XP);
+
+	new = readl(port->base + RP_XP_CTL_1);
+	new &= ~RP_XP_CTL_1_OLD_IOBIST_EN;
+	writel(new, port->base + RP_XP_CTL_1);
+
+	writel(0x10000001, port->base + RP_VEND_XP_BIST);
+	writel(0, port->base + RP_PRBS);
+
+	mdelay(1);
+
+	writel(0x90820001, port->base + RP_VEND_XP_BIST);
+	new = readl(port->base + RP_VEND_XP_BIST);
+
+	new = readl(port->base + RP_XP_CTL_1);
+	new |= RP_XP_CTL_1_OLD_IOBIST_EN;
+	writel(new, port->base + RP_XP_CTL_1);
+
+	new = readl(port->base + RP_VEND_XP);
+	new |= RP_VEND_XP_PRBS_EN;
+	writel(new, port->base + RP_VEND_XP);
+
+	mdelay(1000);
+
+	new = readl(port->base + RP_VEND_XP);
+	port->loopback_stat = (new & RP_VEND_XP_PRBS_STAT) >> 2;
+	pr_info("--- loopback status ---\n");
+	for (i = 0; i < port->lanes; ++i)
+		pr_info("@lane %d: %s\n", i,
+				(port->loopback_stat & 0x01 << i) ? "pass" : "fail");
+
+	new = readl(port->base + RP_PRBS);
+	pr_info("--- PRBS pattern locked ---\n");
+	for (i = 0; i < port->lanes; ++i)
+		pr_info("@lane %d: %s\n", i,
+				(new >> 16 & 0x01 << i) ? "Y" : "N");
+	pr_info("--- err overflow bits ---\n");
+	for (i = 0; i < port->lanes; ++i)
+		pr_info("@lane %d: %s\n", i,
+				((new & 0xffff) & 0x01 << i) ? "Y" : "N");
+
+	new = readl(port->base + RP_XP_CTL_1);
+	new &= ~RP_XP_CTL_1_OLD_IOBIST_EN;
+	writel(new, port->base + RP_XP_CTL_1);
+
+	pr_info("--- err counts ---\n");
+	for (i = 0; i < port->lanes; ++i) {
+		writel(i, port->base + RP_LANE_PRBS_ERR_COUNT);
+		new = readl(port->base + RP_LANE_PRBS_ERR_COUNT);
+		pr_info("@lane %d: %u\n", i, new >> 16);
+	}
+
+	writel(0x90000001, port->base + RP_VEND_XP_BIST);
+
+	new = readl(port->base + RP_VEND_XP);
+	new &= ~RP_VEND_XP_PRBS_EN;
+	writel(new, port->base + RP_VEND_XP);
+
+	mdelay(1);
+
+	writel(0x92000001, port->base + RP_VEND_XP_BIST);
+	writel(0x90000001, port->base + RP_VEND_XP_BIST);
+	pr_info("pcie loopback test is done\n");
+
+	return 0;
+}
+
+static int apply_lane_width(struct seq_file *s, void *data)
+{
+	unsigned int new;
+	struct tegra_pcie_port *port = (struct tegra_pcie_port *)(s->private);
+
+	if (port->lanes > 0x10) {
+		seq_puts(s, "link width cannot be grater than 16\n");
+		new = readl(port->base + RP_LINK_CONTROL_STATUS);
+		port->lanes = (new &
+				RP_LINK_CONTROL_STATUS_NEG_LINK_WIDTH) >> 20;
+		return 0;
+	}
+	new = readl(port->base + RP_VEND_XP1);
+	new &= ~RP_VEND_XP1_RNCTRL_MAXWIDTH_MASK;
+	new |= port->lanes | RP_VEND_XP1_RNCTRL_EN;
+	writel(new, port->base + RP_VEND_XP1);
+	mdelay(1);
+
+	new = readl(port->base + RP_LINK_CONTROL_STATUS);
+	new = (new & RP_LINK_CONTROL_STATUS_NEG_LINK_WIDTH) >> 20;
+	if (new != port->lanes)
+		seq_printf(s, "can't set link width %u, falling back to %u\n",
+				port->lanes, new);
+	else
+		seq_printf(s, "lane width %d applied\n", new);
+	port->lanes = new;
+	return 0;
+}
+
+static int aspm_state_cnt(struct seq_file *s, void *data)
+{
+	u32 val, cs;
+	struct tegra_pcie_port *port = (struct tegra_pcie_port *)(s->private);
+
+	cs = readl(port->base + RP_LINK_CONTROL_STATUS);
+	/* check if L0s is enabled on this port */
+	if (cs & RP_LINK_CONTROL_STATUS_L0s_ENABLED) {
+		val = readl(port->base + RP_PRIV_XP_TX_L0S_ENTRY_COUNT);
+		seq_printf(s, "Tx L0s entry count : %u\n", val);
+	} else {
+		seq_printf(s, "Tx L0s entry count : %s\n", "disabled");
+	}
+
+	val = readl(port->base + RP_PRIV_XP_RX_L0S_ENTRY_COUNT);
+	seq_printf(s, "Rx L0s entry count : %u\n", val);
+
+	/* check if L1 is enabled on this port */
+	if (cs & RP_LINK_CONTROL_STATUS_L1_ENABLED) {
+		val = readl(port->base + RP_PRIV_XP_TX_L1_ENTRY_COUNT);
+		seq_printf(s, "Link L1 entry count : %u\n", val);
+	} else {
+		seq_printf(s, "Link L1 entry count : %s\n", "disabled");
+	}
+
+	cs = readl(port->base + RP_L1_PM_SS_CONTROL);
+	/*
+	 * Resetting the count value is not possible by any means
+	 * because of HW Bug : 200034278
+	 */
+	/* check if L1.1 is enabled */
+	if (cs & RP_L1_PM_SS_CONTROL_ASPM_L11_ENABLE) {
+		val = readl(port->base + RP_L1_1_ENTRY_COUNT);
+		val |= RP_L1_1_ENTRY_COUNT_RESET;
+		writel(val, port->base + RP_L1_1_ENTRY_COUNT);
+		seq_printf(s, "Link L1.1 entry count : %u\n", (val & 0xFFFF));
+	} else {
+		seq_printf(s, "Link L1.1 entry count : %s\n", "disabled");
+	}
+	/* check if L1.2 is enabled */
+	if (cs & RP_L1_PM_SS_CONTROL_ASPM_L12_ENABLE) {
+		val = readl(port->base + RP_L1_2_ENTRY_COUNT);
+		val |= RP_L1_2_ENTRY_COUNT_RESET;
+		writel(val, port->base + RP_L1_2_ENTRY_COUNT);
+		seq_printf(s, "Link L1.2 entry count : %u\n", (val & 0xFFFF));
+	} else {
+		seq_printf(s, "Link L1.2 entry count : %s\n", "disabled");
+	}
+
+	return 0;
+}
+
+static char *aspm_states[] = {
+	"Tx-L0s",
+	"Rx-L0s",
+	"L1",
+	"IDLE ((Tx-L0s && Rx-L0s) + L1)"
+};
+
+static int list_aspm_states(struct seq_file *s, void *data)
+{
+	u32 i = 0;
+
+	seq_puts(s, "----------------------------------------------------\n");
+	seq_puts(s, "Note: Duration of link's residency is calcualated\n");
+	seq_puts(s, "      only for one of the ASPM states at a time\n");
+	seq_puts(s, "----------------------------------------------------\n");
+	seq_puts(s, "write(echo) number from below table corresponding to\n");
+	seq_puts(s, "one of the ASPM states for which link duration needs\n");
+	seq_puts(s, "to be calculated to 'config_aspm_state'\n");
+	seq_puts(s, "-----------------\n");
+	for (i = 0; i < ARRAY_SIZE(aspm_states); i++)
+		seq_printf(s, "%d : %s\n", i, aspm_states[i]);
+	seq_puts(s, "-----------------\n");
+	return 0;
+}
+
+static int apply_aspm_state(struct seq_file *s, void *data)
+{
+	u32 val;
+	struct tegra_pcie_port *port = (struct tegra_pcie_port *)(s->private);
+
+	if (config_aspm_state >= ARRAY_SIZE(aspm_states)) {
+		seq_printf(s, "Invalid ASPM state : %u\n", config_aspm_state);
+		list_aspm_states(s, data);
+	} else {
+		val = readl(port->base + RP_PRIV_XP_CONFIG);
+		val &= ~RP_PRIV_XP_CONFIG_LOW_PWR_DURATION_MASK;
+		val |= config_aspm_state;
+		writel(val, port->base + RP_PRIV_XP_CONFIG);
+		seq_printf(s, "Configured for ASPM-%s state...\n",
+				aspm_states[config_aspm_state]);
+	}
+	return 0;
+}
+
+static int get_aspm_duration(struct seq_file *s, void *data)
+{
+	u32 val;
+	struct tegra_pcie_port *port = (struct tegra_pcie_port *)(s->private);
+
+	val = readl(port->base + RP_PRIV_XP_DURATION_IN_LOW_PWR_100NS);
+	/* 52.08 = 1000 / 19.2MHz is rounded to 52      */
+	seq_printf(s, "ASPM-%s duration = %d ns\n",
+			aspm_states[config_aspm_state], (u32)((val * 100)/52));
+	return 0;
+}
+
+static int secondary_bus_reset(struct seq_file *s, void *data)
+{
+	u32 val;
+	struct tegra_pcie_port *port = (struct tegra_pcie_port *)(s->private);
+
+	val = readl(port->base + RP_INTR_BCR);
+	val |= RP_INTR_BCR_SB_RESET;
+	writel(val, port->base + RP_INTR_BCR);
+	usleep_range(10, 11);
+	val = readl(port->base + RP_INTR_BCR);
+	val &= ~RP_INTR_BCR_SB_RESET;
+	writel(val, port->base + RP_INTR_BCR);
+
+	seq_puts(s, "Secondary Bus Reset applied successfully...\n");
+	return 0;
+}
+
+static void reset_l1ss_counter(struct tegra_pcie_port *port, u32 val,
+		unsigned long offset)
+{
+	int c = 0;
+
+	if ((val & 0xFFFF) == 0xFFFF) {
+		pr_info(" Trying reset L1ss entry count to 0\n");
+		while (val) {
+			if (c++ > 50) {
+				pr_info("Timeout: reset did not happen!\n");
+				break;
+			}
+			val |= RP_L1_1_ENTRY_COUNT_RESET;
+			writel(val, port->base + offset);
+			mdelay(1);
+			val = readl(port->base + offset);
+		}
+		if (!val)
+			pr_info("L1ss entry count reset to 0\n");
+	}
+}
+
+static int aspm_l11(struct seq_file *s, void *data)
+{
+	struct pci_dev *pdev = NULL;
+	u32 val = 0, pos = 0;
+	struct tegra_pcie_port *port = NULL;
+	struct tegra_pcie *pcie = (struct tegra_pcie *)(s->private);
+
+	pr_info("\nPCIE aspm l1.1 test START..\n");
+	list_for_each_entry(port, &pcie->ports, list) {
+		/* reset RP L1.1 counter */
+		val = readl(port->base + RP_L1_1_ENTRY_COUNT);
+		val |= RP_L1_1_ENTRY_COUNT_RESET;
+		writel(val, port->base + RP_L1_1_ENTRY_COUNT);
+
+		val = readl(port->base + RP_L1_1_ENTRY_COUNT);
+		pr_info("L1.1 Entry count before %x\n", val);
+		reset_l1ss_counter(port, val, RP_L1_1_ENTRY_COUNT);
+	}
+	/* disable automatic l1ss exit by gpu */
+	for_each_pci_dev(pdev)
+		if (pci_pcie_type(pdev) != PCI_EXP_TYPE_ROOT_PORT) {
+			pci_write_config_dword(pdev, 0x658, 0);
+			pci_write_config_dword(pdev, 0x150, 0xE0000015);
+		}
+	for_each_pci_dev(pdev) {
+		u16 aspm;
+
+		pcie_capability_read_word(pdev, PCI_EXP_LNKCTL, &aspm);
+		aspm |= PCI_EXP_LNKCTL_ASPM_L1;
+		pcie_capability_write_word(pdev, PCI_EXP_LNKCTL, aspm);
+		pos = pci_find_ext_capability(pdev, PCI_EXT_CAP_ID_L1SS);
+		pci_read_config_dword(pdev, pos + PCI_L1SS_CTL1, &val);
+		val &= ~PCI_L1SS_CTL1_L1SS_MASK;
+		val |= PCI_L1SS_CTL1_ASPM_L1_1;
+		pci_write_config_dword(pdev, pos + PCI_L1SS_CTL1, val);
+		if (pci_pcie_type(pdev) != PCI_EXP_TYPE_ROOT_PORT)
+			break;
+	}
+	mdelay(2000);
+	for_each_pci_dev(pdev) {
+		pos = pci_find_ext_capability(pdev, PCI_EXT_CAP_ID_L1SS);
+		pci_read_config_dword(pdev, pos + PCI_L1SS_CTL1, &val);
+		if (pci_pcie_type(pdev) != PCI_EXP_TYPE_ROOT_PORT)
+			break;
+	}
+	list_for_each_entry(port, &pcie->ports, list) {
+		val = readl(port->base + RP_L1_1_ENTRY_COUNT);
+		pr_info("L1.1 Entry count after %x\n", val);
+	}
+
+	pr_info("PCIE aspm l1.1 test END..\n");
+	return 0;
+}
+
+static int aspm_l1ss(struct seq_file *s, void *data)
+{
+	struct pci_dev *pdev = NULL;
+	u32 val = 0, pos = 0;
+	struct tegra_pcie_port *port = NULL;
+	struct tegra_pcie *pcie = (struct tegra_pcie *)(s->private);
+
+	pr_info("\nPCIE aspm l1ss test START..\n");
+	list_for_each_entry(port, &pcie->ports, list) {
+		/* reset RP L1.1 L1.2 counters */
+		val = readl(port->base + RP_L1_1_ENTRY_COUNT);
+		val |= RP_L1_1_ENTRY_COUNT_RESET;
+		writel(val, port->base + RP_L1_1_ENTRY_COUNT);
+		val = readl(port->base + RP_L1_1_ENTRY_COUNT);
+		pr_info("L1.1 Entry count before %x\n", val);
+		reset_l1ss_counter(port, val, RP_L1_1_ENTRY_COUNT);
+
+		val = readl(port->base + RP_L1_2_ENTRY_COUNT);
+		val |= RP_L1_2_ENTRY_COUNT_RESET;
+		writel(val, port->base + RP_L1_2_ENTRY_COUNT);
+		val = readl(port->base + RP_L1_2_ENTRY_COUNT);
+		pr_info("L1.2 Entry count before %x\n", val);
+		reset_l1ss_counter(port, val, RP_L1_2_ENTRY_COUNT);
+	}
+	/* disable automatic l1ss exit by gpu */
+	for_each_pci_dev(pdev)
+		if (pci_pcie_type(pdev) != PCI_EXP_TYPE_ROOT_PORT) {
+			pci_write_config_dword(pdev, 0x658, 0);
+			pci_write_config_dword(pdev, 0x150, 0xE0000015);
+		}
+
+	for_each_pci_dev(pdev) {
+		u16 aspm;
+
+		pcie_capability_read_word(pdev, PCI_EXP_LNKCTL, &aspm);
+		aspm |= PCI_EXP_LNKCTL_ASPM_L1;
+		pcie_capability_write_word(pdev, PCI_EXP_LNKCTL, aspm);
+		pos = pci_find_ext_capability(pdev, PCI_EXT_CAP_ID_L1SS);
+		pci_read_config_dword(pdev, pos + PCI_L1SS_CTL1, &val);
+		val &= ~PCI_L1SS_CTL1_L1SS_MASK;
+		val |= (PCI_L1SS_CTL1_ASPM_L1_1 | PCI_L1SS_CTL1_ASPM_L1_2);
+		pci_write_config_dword(pdev, pos + PCI_L1SS_CTL1, val);
+		if (pci_pcie_type(pdev) != PCI_EXP_TYPE_ROOT_PORT)
+			break;
+	}
+	mdelay(2000);
+	for_each_pci_dev(pdev) {
+		pos = pci_find_ext_capability(pdev, PCI_EXT_CAP_ID_L1SS);
+		pci_read_config_dword(pdev, pos + PCI_L1SS_CTL1, &val);
+		if (pci_pcie_type(pdev) != PCI_EXP_TYPE_ROOT_PORT)
+			break;
+	}
+	list_for_each_entry(port, &pcie->ports, list) {
+		u32 ltr_val;
+
+		val = readl(port->base + RP_L1_1_ENTRY_COUNT);
+		pr_info("L1.1 Entry count after %x\n", val);
+		val = readl(port->base + RP_L1_2_ENTRY_COUNT);
+		pr_info("L1.2 Entry count after %x\n", val);
+
+		val = readl(port->base + RP_LTR_REP_VAL);
+		pr_info("LTR reproted by EP %x\n", val);
+		ltr_val = (val & 0x1FF) * (1 << (5 * ((val & 0x1C00) >> 10)));
+		if (ltr_val > (106 * 1000)) {
+			pr_info("EP's LTR = %u ns is > RP's threshold = %u ns\n",
+					ltr_val, 106 * 1000);
+			pr_info("Hence only L1.2 entry allowed\n");
+		} else {
+			pr_info("EP's LTR = %u ns is < RP's threshold = %u ns\n",
+					ltr_val, 106 * 1000);
+			pr_info("Hence only L1.1 entry allowed\n");
+		}
+	}
+
+	pr_info("PCIE aspm l1ss test END..\n");
+	return 0;
+}
+
+struct ltssm_major_state {
+	const char *name;
+	const char *minor[8];
+};
+
+struct ltssm_state {
+	struct ltssm_major_state major[12];
+};
+
+static struct ltssm_state ltssm_state = {
+	.major[0]  = {"detect", {"quiet", "active", "retry", "wait", "entry"} },
+	.major[1]  = {"polling", {"active", "config", "idle", NULL,
+		"compliance", "cspeed"} },
+	.major[2]  = {"config", {"link start", "link accept", "lane accept",
+		"lane wait", "idle", "pwrup", "complete"} },
+	.major[3]  = {NULL, {NULL} },
+	.major[4]  = {"l0", {"normal", "l0s entry", "l0s idle", "l0s wait",
+		"l0s fts", "pwrup"} },
+	.major[5]  = {"l1", {"entry", "waitrx", "idle", "wait", "pwrup",
+		"beacon entry", "beacon exit"} },
+	.major[6]  = {"l2", {"entry", "waitrx", "transmitwake", "idle"} },
+	.major[7]  = {"recovery", {"rcvrlock", "rcvrcfg", "speed", "idle",
+		NULL, NULL, NULL, "finish pkt"} },
+	.major[8]  = {"loopback", {"entry", "active", "idle", "exit", "speed",
+		"pre speed"} },
+	.major[9]  = {"hotreset", {NULL} },
+	.major[10] = {"disabled", {NULL} },
+	.major[11] = {"txchar", {NULL} },
+};
+
+static const char *ltssm_get_major(unsigned int major)
+{
+	const char *state;
+
+	state = ltssm_state.major[major].name;
+	if (!state)
+		return "unknown";
+	return state;
+}
+
+static const char *ltssm_get_minor(unsigned int major, unsigned int minor)
+{
+	const char *state;
+
+	state = ltssm_state.major[major].minor[minor];
+	if (!state)
+		return "unknown";
+	return state;
+}
+
+static int dump_ltssm_trace(struct seq_file *s, void *data)
+{
+	struct tegra_pcie_port *port = (struct tegra_pcie_port *)(s->private);
+	unsigned int val, ridx, widx, entries;
+
+	seq_puts(s, "LTSSM trace dump:\n");
+	val = readl(port->base + RP_LTSSM_TRACE_STATUS);
+	widx = LTSSM_TRACE_STATUS_WRITE_POINTER(val);
+	entries = LTSSM_TRACE_STATUS_RAM_FULL(val) ? 32 : widx;
+	seq_printf(s, "LTSSM trace dump - %d entries:\n", entries);
+	for (ridx = 0; ridx < entries; ridx++) {
+		val = LTSSM_TRACE_STATUS_READ_ADDR(ridx);
+		writel(val, port->base + RP_LTSSM_TRACE_STATUS);
+		val = readl(port->base + RP_LTSSM_TRACE_STATUS);
+
+		seq_printf(s, "[0x%08x] major: %-10s minor_tx: %-15s minor_rx: %s\n",
+				val,
+				ltssm_get_major(LTSSM_TRACE_STATUS_MAJOR(val)),
+				ltssm_get_minor(LTSSM_TRACE_STATUS_MAJOR(val),
+					LTSSM_TRACE_STATUS_PTX_MINOR(val)),
+				ltssm_get_minor(LTSSM_TRACE_STATUS_MAJOR(val),
+					LTSSM_TRACE_STATUS_PRX_MINOR(val)));
+	}
+	/* clear trace ram */
+	val = readl(port->base + RP_LTSSM_TRACE_CONTROL);
+	val |= LTSSM_TRACE_CONTROL_CLEAR_RAM;
+	writel(val, port->base + RP_LTSSM_TRACE_CONTROL);
+	val &= ~LTSSM_TRACE_CONTROL_CLEAR_RAM;
+	writel(val, port->base + RP_LTSSM_TRACE_CONTROL);
+
+	return 0;
+}
+
+static struct dentry *create_tegra_pcie_debufs_file(char *name,
+		const struct file_operations *ops,
+		struct dentry *parent,
+		void *data)
+{
+	struct dentry *d;
+
+	d = debugfs_create_file(name, 0444, parent, data, ops);
+	if (!d)
+		debugfs_remove_recursive(parent);
+
+	return d;
+}
+
+#define DEFINE_ENTRY(__name)	\
+static int __name ## _open(struct inode *inode, struct file *file)	\
+{									\
+	return single_open(file, __name, inode->i_private);		\
+}									\
+static const struct file_operations __name ## _fops = {		\
+	.open		= __name ## _open,			\
+	.read		= seq_read,				\
+	.llseek		= seq_lseek,				\
+	.release	= single_release,			\
+};
+
+/* common */
+DEFINE_ENTRY(list_devices)
+DEFINE_ENTRY(apply_link_speed)
+DEFINE_ENTRY(check_d3hot)
+DEFINE_ENTRY(dump_config_space)
+DEFINE_ENTRY(dump_afi_space)
+DEFINE_ENTRY(config_read)
+DEFINE_ENTRY(config_write)
+DEFINE_ENTRY(aspm_l11)
+DEFINE_ENTRY(aspm_l1ss)
+DEFINE_ENTRY(power_down)
+
+/* Port specific */
+DEFINE_ENTRY(loopback)
+DEFINE_ENTRY(apply_lane_width)
+DEFINE_ENTRY(aspm_state_cnt)
+DEFINE_ENTRY(list_aspm_states)
+DEFINE_ENTRY(apply_aspm_state)
+DEFINE_ENTRY(get_aspm_duration)
+DEFINE_ENTRY(secondary_bus_reset)
+DEFINE_ENTRY(dump_ltssm_trace)
+
+static int tegra_pcie_port_debugfs_init(struct tegra_pcie_port *port)
+{
+	struct dentry *d;
+	char port_name[2] = {0};
+
+	snprintf(port_name, sizeof(port_name), "%d", port->index);
+	port->port_debugfs = debugfs_create_dir(port_name,
+			port->pcie->debugfs);
+	if (!port->port_debugfs)
+		return -ENOMEM;
+
+	debugfs_create_u32("lane_width", 0664, port->port_debugfs,
+			&(port->lanes));
+	debugfs_create_x32("loopback_status", 0664, port->port_debugfs,
+			&(port->loopback_stat));
+
+	d = debugfs_create_file("loopback", 0444, port->port_debugfs,
+			(void *)port, &loopback_fops);
+	if (!d)
+		goto remove;
+
+	d = debugfs_create_file("apply_lane_width", 0444, port->port_debugfs,
+			(void *)port, &apply_lane_width_fops);
+	if (!d)
+		goto remove;
+
+	d = debugfs_create_file("aspm_state_cnt", 0444, port->port_debugfs,
+			(void *)port, &aspm_state_cnt_fops);
+	if (!d)
+		goto remove;
+
+	debugfs_create_u16("config_aspm_state", 0664, port->port_debugfs,
+			&config_aspm_state);
+
+	d = debugfs_create_file("apply_aspm_state", 0444, port->port_debugfs,
+			(void *)port, &apply_aspm_state_fops);
+	if (!d)
+		goto remove;
+
+	d = debugfs_create_file("list_aspm_states", 0444, port->port_debugfs,
+			(void *)port, &list_aspm_states_fops);
+	if (!d)
+		goto remove;
+
+	d = debugfs_create_file("dump_ltssm_trace", 0444, port->port_debugfs,
+			(void *)port, &dump_ltssm_trace_fops);
+	if (!d)
+		goto remove;
+
+	d = debugfs_create_file("get_aspm_duration", 0444, port->port_debugfs,
+			(void *)port, &get_aspm_duration_fops);
+	if (!d)
+		goto remove;
+
+	d = debugfs_create_file("secondary_bus_reset", 0444, port->port_debugfs,
+			(void *)port, &secondary_bus_reset_fops);
+	if (!d)
+		goto remove;
+
+	return 0;
+
+remove:
+	debugfs_remove_recursive(port->port_debugfs);
+	port->port_debugfs = NULL;
+	return -ENOMEM;
+}
+
 static void *tegra_pcie_ports_seq_start(struct seq_file *s, loff_t *pos)
 {
 	struct tegra_pcie *pcie = s->private;
@@ -2609,12 +4228,91 @@ static void tegra_pcie_debugfs_exit(stru
 	pcie->debugfs = NULL;
 }
 
-static void tegra_pcie_debugfs_init(struct tegra_pcie *pcie)
+static int tegra_pcie_debugfs_init(struct tegra_pcie *pcie)
 {
+	struct dentry *d;
+	struct tegra_pcie_port *port;
+
 	pcie->debugfs = debugfs_create_dir("pcie", NULL);
 
 	debugfs_create_file("ports", S_IFREG | S_IRUGO, pcie->debugfs, pcie,
 			    &tegra_pcie_ports_fops);
+
+	d = create_tegra_pcie_debufs_file("list_devices",
+			&list_devices_fops, pcie->debugfs,
+			(void *)pcie);
+	if (!d)
+		goto remove;
+
+	d = debugfs_create_bool("is_gen2_speed(WO)", 0200, pcie->debugfs,
+			&is_gen2_speed);
+	if (!d)
+		goto remove;
+
+	d = create_tegra_pcie_debufs_file("apply_link_speed",
+			&apply_link_speed_fops, pcie->debugfs,
+			(void *)pcie);
+	if (!d)
+		goto remove;
+
+	d = create_tegra_pcie_debufs_file("check_d3hot",
+			&check_d3hot_fops, pcie->debugfs,
+			(void *)pcie);
+	if (!d)
+		goto remove;
+
+	d = create_tegra_pcie_debufs_file("power_down",
+			&power_down_fops, pcie->debugfs,
+			(void *)pcie);
+	if (!d)
+		goto remove;
+
+	d = create_tegra_pcie_debufs_file("dump_config_space",
+			&dump_config_space_fops,
+			pcie->debugfs, (void *)pcie);
+	if (!d)
+		goto remove;
+
+	d = create_tegra_pcie_debufs_file("dump_afi_space",
+			&dump_afi_space_fops, pcie->debugfs,
+			(void *)pcie);
+	if (!d)
+		goto remove;
+
+	debugfs_create_u16("bus_dev_func", 0664, pcie->debugfs, &bdf);
+	debugfs_create_u16("config_offset", 0664, pcie->debugfs,
+			&config_offset);
+
+	debugfs_create_u32("config_val", 0664, pcie->debugfs, &config_val);
+
+	d = create_tegra_pcie_debufs_file("config_read", &config_read_fops,
+			pcie->debugfs, (void *)pcie);
+	if (!d)
+		goto remove;
+
+	d = create_tegra_pcie_debufs_file("config_write", &config_write_fops,
+			pcie->debugfs, (void *)pcie);
+	if (!d)
+		goto remove;
+	d = create_tegra_pcie_debufs_file("aspm_l11", &aspm_l11_fops,
+			pcie->debugfs, (void *)pcie);
+	if (!d)
+		goto remove;
+	d = create_tegra_pcie_debufs_file("aspm_l1ss", &aspm_l1ss_fops,
+			pcie->debugfs, (void *)pcie);
+	if (!d)
+		goto remove;
+
+	list_for_each_entry(port, &pcie->ports, list) {
+		if (tegra_pcie_port_debugfs_init(port))
+			goto remove;
+	}
+
+	return 0;
+
+remove:
+	tegra_pcie_debugfs_exit(pcie);
+	return -ENOMEM;
 }
 
 static int tegra_pcie_probe(struct platform_device *pdev)
@@ -2622,6 +4320,7 @@ static int tegra_pcie_probe(struct platf
 	struct device *dev = &pdev->dev;
 	struct pci_host_bridge *host;
 	struct tegra_pcie *pcie;
+	struct tegra_pcie_port *port = NULL;
 	int err;
 
 	host = devm_pci_alloc_host_bridge(dev, sizeof(*pcie));
@@ -2636,6 +4335,12 @@ static int tegra_pcie_probe(struct platf
 	INIT_LIST_HEAD(&pcie->ports);
 	pcie->dev = dev;
 
+#if IS_ENABLED(CONFIG_TEGRA_BWMGR)
+	pcie->emc_bwmgr = tegra_bwmgr_register(TEGRA_BWMGR_CLIENT_PCIE);
+	if (!pcie->emc_bwmgr)
+		dev_err(dev, "couldn't register with EMC BwMgr\n");
+#endif
+
 	err = tegra_pcie_parse_dt(pcie);
 	if (err < 0)
 		return err;
@@ -2652,6 +4357,11 @@ static int tegra_pcie_probe(struct platf
 		goto put_resources;
 	}
 
+	list_for_each_entry(port, &pcie->ports, list)
+		if (port->has_mxm_port && tegra_pcie_mxm_pwr_init(port))
+			dev_info(dev, "pwr_good is down for port %d, ignoring\n",
+				 port->index);
+
 	pm_runtime_enable(pcie->dev);
 	err = pm_runtime_get_sync(pcie->dev);
 	if (err < 0) {
@@ -2659,6 +4369,27 @@ static int tegra_pcie_probe(struct platf
 		goto pm_runtime_put;
 	}
 
+	/*
+	 * If all PCIe ports are down, power gate PCIe. This can happen if
+	 * no endpoints are connected, so don't fail probe.
+	 */
+	port = NULL;
+	err = -ENOMEDIUM;
+
+	list_for_each_entry(port, &pcie->ports, list) {
+		if (tegra_pcie_link_up(port)) {
+			err = 0;
+			break;
+		}
+	}
+
+	if (err == -ENOMEDIUM) {
+		err = 0;
+		goto pm_runtime_put;
+	}
+
+	pci_add_flags(PCI_REASSIGN_ALL_BUS);
+
 	host->ops = &tegra_pcie_ops;
 	host->map_irq = tegra_pcie_map_irq;
 
@@ -2671,6 +4402,8 @@ static int tegra_pcie_probe(struct platf
 	if (IS_ENABLED(CONFIG_DEBUG_FS))
 		tegra_pcie_debugfs_init(pcie);
 
+	device_init_wakeup(dev, true);
+
 	return 0;
 
 pm_runtime_put:
@@ -2688,9 +4421,19 @@ static int tegra_pcie_remove(struct plat
 	struct pci_host_bridge *host = pci_host_bridge_from_priv(pcie);
 	struct tegra_pcie_port *port, *tmp;
 
+	if (list_empty(&pcie->ports))
+		return 0;
+
+	device_init_wakeup(&pdev->dev, false);
+
 	if (IS_ENABLED(CONFIG_DEBUG_FS))
 		tegra_pcie_debugfs_exit(pcie);
 
+#if IS_ENABLED(CONFIG_TEGRA_BWMGR)
+	tegra_bwmgr_set_emc(pcie->emc_bwmgr, 0, TEGRA_BWMGR_SET_EMC_FLOOR);
+	tegra_bwmgr_unregister(pcie->emc_bwmgr);
+#endif
+
 	pci_stop_root_bus(host->bus);
 	pci_remove_root_bus(host->bus);
 	pm_runtime_put_sync(pcie->dev);
@@ -2713,6 +4456,9 @@ static int __maybe_unused tegra_pcie_pm_
 	struct tegra_pcie_port *port;
 	int err;
 
+	if (list_empty(&pcie->ports))
+		return 0;
+
 	list_for_each_entry(port, &pcie->ports, list)
 		tegra_pcie_pme_turnoff(port);
 
@@ -2738,6 +4484,7 @@ static int __maybe_unused tegra_pcie_pm_
 
 	pinctrl_pm_select_idle_state(dev);
 	tegra_pcie_power_off(pcie);
+	tegra_pcie_config_plat(pcie, 0);
 
 	return 0;
 }
@@ -2747,6 +4494,11 @@ static int __maybe_unused tegra_pcie_pm_
 	struct tegra_pcie *pcie = dev_get_drvdata(dev);
 	int err;
 
+	if (list_empty(&pcie->ports))
+		return 0;
+
+	tegra_pcie_config_plat(pcie, 1);
+
 	err = tegra_pcie_power_on(pcie);
 	if (err) {
 		dev_err(dev, "tegra pcie power on fail: %d\n", err);
@@ -2797,10 +4549,38 @@ poweroff:
 	return err;
 }
 
+static int tegra_pcie_suspend_late(struct device *dev)
+{
+	struct tegra_pcie *pcie = dev_get_drvdata(dev);
+
+	if (list_empty(&pcie->ports))
+		return 0;
+
+	if (gpio_is_valid(pcie->pex_wake))
+		enable_irq_wake(gpio_to_irq(pcie->pex_wake));
+
+	return 0;
+}
+
+static int tegra_pcie_resume_early(struct device *dev)
+{
+	struct tegra_pcie *pcie = dev_get_drvdata(dev);
+
+	if (list_empty(&pcie->ports))
+		return 0;
+
+	if (gpio_is_valid(pcie->pex_wake))
+		disable_irq_wake(gpio_to_irq(pcie->pex_wake));
+
+	return 0;
+}
+
 static const struct dev_pm_ops tegra_pcie_pm_ops = {
 	SET_RUNTIME_PM_OPS(tegra_pcie_pm_suspend, tegra_pcie_pm_resume, NULL)
 	SET_NOIRQ_SYSTEM_SLEEP_PM_OPS(tegra_pcie_pm_suspend,
 				      tegra_pcie_pm_resume)
+	SET_LATE_SYSTEM_SLEEP_PM_OPS(tegra_pcie_suspend_late,
+				     tegra_pcie_resume_early)
 };
 
 static struct platform_driver tegra_pcie_driver = {
--- a/drivers/pci/controller/pci-tegra.c
+++ b/drivers/pci/controller/pci-tegra.c
@@ -405,7 +405,6 @@
 #define LINK_RETRAIN_TIMEOUT 100000 /* in usec */
 
 struct tegra_msi {
-	struct msi_controller chip;
 	DECLARE_BITMAP(used, INT_PCI_MSI_NR);
 	struct irq_domain *domain;
 	struct mutex map_lock;
@@ -4244,10 +4243,8 @@ static int tegra_pcie_debugfs_init(struc
 	if (!d)
 		goto remove;
 
-	d = debugfs_create_bool("is_gen2_speed(WO)", 0200, pcie->debugfs,
+	debugfs_create_bool("is_gen2_speed(WO)", 0200, pcie->debugfs,
 			&is_gen2_speed);
-	if (!d)
-		goto remove;
 
 	d = create_tegra_pcie_debufs_file("apply_link_speed",
 			&apply_link_speed_fops, pcie->debugfs,
